### PR TITLE
[DiscardingTG] Undo pointer auth workaround; fix memory leaks

### PR DESCRIPTION
**Pull request by ktoso**
There are two leaks in the group.

"first error task leaks"
(a) the "the first error task that we RETAIN" and later on use to automatically re-throw the "first thrown by a child task error" task was not released;

"all values leak"
(b) due to the "hot fix" for pointer auth https://github.com/apple/swift/pull/65220 done for rdar://107574868 where we changed the signature of addTask from `()->()` (what it must be),  to `()->T` in order to work around pointer-auth issues we introduced a leak and ALL "discarded" values would be leaking since we would not release that T value. The correct way to solve this is to undo this workaround and use correct pointer auth, which this PR does.

We must fix the method signature and undo the hack which https://github.com/apple/swift/pull/65223 was.

Resolves: rdar://108829466

### CODE REVIEW

File: test/Concurrency/Runtime/async_taskgroup_discarding_noLeaks.swift [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+func shouldEqual<T: Equatable>(_ lhs: T, _ rhs: T) {
+  precondition(lhs == rhs, "'\(lhs)' did not equal '\(rhs)'")
+}
+func shouldStartWith(_ lhs: Any, _ rhs: Any) {
+  let l = "\(lhs)"
+  let r = "\(rhs)"
+  precondition(l.prefix("\(r)".count) == r, "'\(l)' did not start with '\(r)'")
+}
+
+final class SomeClass: @unchecked Sendable {
```
Comment (ktoso) [inline test/Concurrency/Runtime/async_taskgroup_discarding_noLeaks.swift:43]: FIXME: this still will be leaked if it is a class; something is wrong with how we retain errors in `fillGroupNextErrorResult` 🤔 If we would not retain them there we crash though, so something is off and I don't quite get what yet.

File: test/Concurrency/Runtime/async_taskgroup_discarding_noLeaks.swift [hunk 2]
```diff
@@ -0,0 +1,119 @@
+// RUN: %target-run-simple-swift( -Xfrontend -disable-availability-checking -parse-as-library)
```
Comment (ktoso) [inline test/Concurrency/Runtime/async_taskgroup_discarding_noLeaks.swift:1]: TODO: we need this to run with leak detection

File: stdlib/public/Concurrency/TaskGroup.cpp [hunk 3]
```diff
@@ -1040,7 +1040,7 @@ static void fillGroupNextResult(TaskFutureWaitAsyncContext *context,
 
   case PollStatus::Error: {
     auto error = reinterpret_cast<SwiftError *>(result.storage);
-    fillGroupNextErrorResult(context, error);
+    fillGroupNextErrorResult(context, error); // FIXME: this specifically retains the error, but likely should not!??!!?
```
[the remark this answers is not in the document]
Comment (ktoso) [reply stdlib/public/Concurrency/TaskGroup.cpp:1043]: This SPECIFICALLY does
```
  void fillWithError(SwiftError *error) {
    errorResult = error;
    swift_errorRetain(error);
  }
```

and if we don't retain it we get crashes... so... how and where are we missing a balanced release, I'm a bit confused 🤔

File: test/Concurrency/Runtime/async_taskgroup_discarding_noLeaks.swift (hunk 2 above)
Comment (al45tair) [reply test/Concurrency/Runtime/async_taskgroup_discarding_noLeaks.swift:1]: I think you'd just add `xcrun leaks -atExit --` to the front of this command, and make sure that this test only runs on macOS.

Comment (ktoso) [reply test/Concurrency/Runtime/async_taskgroup_discarding_noLeaks.swift]: Yeap, needed to add a new lit target but that sounds like a good idea anyway: https://github.com/apple/swift/pull/65613/commits/5ca9af6a0401418f412fb8b5cd3ccca7979e564f

And managed to fix the leak; though one weird edge case about `Error` remains

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: stdlib/public/Concurrency/DiscardingTaskGroup.swift [hunk 4]
```diff
@@ -171,21 +171,21 @@ public struct DiscardingTaskGroup {
   #if SWIFT_STDLIB_TASK_TO_THREAD_MODEL_CONCURRENCY
   @available(*, unavailable, message: "Unavailable in task-to-thread concurrency model", renamed: "addTask(operation:)")
   #endif
-  public mutating func addTask<DiscardedResult>(
+  public mutating func addTask(
     priority: TaskPriority? = nil,
-    operation: __owned @Sendable @escaping () async -> DiscardedResult
```
Comment (ktoso) [inline stdlib/public/Concurrency/DiscardingTaskGroup.swift:176]: This entire change to `() -> T` was a hotfix to temporarily unbreak but it is very wrong, we must get back to -> Void

File: stdlib/public/Concurrency/Task.cpp [hunk 5]
```diff
@@ -1060,19 +1061,30 @@ void swift::swift_task_run_inline(OpaqueValue *result, void *closureAFP,
...
-          SpecialPointerAuthDiscriminators::AsyncFutureFunction>(closureEntry);
+  TaskCreateFlags taskCreateFlags(rawTaskCreateFlags);
+
+  if (taskCreateFlags.isDiscardingTask()) {
+    std::tie(taskEntry, initialContextSize) =
+        getAsyncClosureEntryPointAndContextSize<
+        FutureAsyncSignature,
+        SpecialPointerAuthDiscriminators::AsyncFutureFunction>(closureEntry);
+//            VoidAsyncSignature,
+//            SpecialPointerAuthDiscriminators::AsyncVoidFunction>(closureEntry);
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.cpp:1078]: Could not figure out yet how to get using these Void ones to compile...

Using them causes:
```
Assertion failed: (I < getNumParams() && "parameter index out of range"), function getExtParameterInfo, file Type.h, line 4571.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/llvm-macosx-arm64/./bin/clang++ -x c++ -c /Users/ktoso/code/swift-project/swift/stdlib/public/Concurrency/Task.cpp -o stdlib/public/Concurrency/CMakeFiles/swift_Concurrency-macosx-arm64.dir/Task.cpp.o -I/Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/stdlib/public/Concurrency -I/Users/ktoso/code/swift-project/swift/stdlib/public/Concurrency -I/Users/ktoso/code/swift-project/swift/stdlib/include -I/Users/ktoso/code/swift-project/swift/stdlib/public/SwiftShims -I/Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/swift-macosx-arm64/include -I/Users/ktoso/code/swift-project/swift/include -I/Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/llvm-macosx-arm64/include -I/Users/ktoso/code/swift-project/llvm-project/clang/include -I/Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/llvm-macosx-arm64/tools/clang/include -I/Users/ktoso/code/swift-project/cmark/src/include -I/Users/ktoso/code/swift-project/build/Ninja-RelWithDebInfoAssert/cmark-macosx-arm64/src -isysroot /Applications/XcodeSEIN.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -I/Users/ktoso/code/swift-project/swift/stdlib/include -isysroot /Applications/XcodeSEIN.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -F/Applications/XcodeSEIN.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/../../../Developer/Library/Frameworks -DCMARK_STATIC_DEFINE -DGTEST_HAS_RTTI=0 -DLLVM_DISABLE_ABI_BREAKING_CHECKS_ENFORCING=1 -DSWIFT_INLINE_NAMESPACE=__runtime -D_DEBUG -D_GLIBCXX_ASSERTIONS -D_LIBCPP_ENABLE_ASSERTIONS -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Dswift_Concurrency_macosx_arm64_EXPORTS -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-class-memaccess -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -Werror=switch -Wdocumentation -Wimplicit-fallthrough -Wunreachable-code -Woverloaded-virtual -DOBJC_OLD_DISPATCH_PROTOTYPES=0 -O2 -DNDEBUG -arch arm64 -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Werror=gnu -Werror=c++98-compat-extra-semi -UNDEBUG -fno-sanitize=all -Dswift_Concurrency_EXPORTS -fswift-async-fp=always -D__STDC_WANT_LIB_EXT1__=1 -DSWIFT_TARGET_LIBRARY_NAME=swift_Concurrency -DSWIFT_RUNTIME -DSWIFT_LIB_SUBDIR=\"macosx\" -DSWIFT_ARCH=\"arm64\" -target arm64-apple-macosx10.13 -O2 -g -DSWIFT_ENABLE_RUNTIME_FUNCTION_COUNTERS -DSWIFT_LIBRARY_EVOLUTION=1 -DSWIFT_STDLIB_SUPPORT_BACK_DEPLOYMENT -DSWIFT_ENABLE_REFLECTION -DSWIFT_STDLIB_HAS_DLADDR -DSWIFT_STDLIB_HAS_DARWIN_LIBMALLOC=1 -DSWIFT_STDLIB_HAS_ASL -DSWIFT_STDLIB_HAS_STDIN -DSWIFT_STDLIB_HAS_ENVIRON -DSWIFT_STDLIB_HAS_LOCALE -DSWIFT_THREADING_DARWIN -DSWIFT_RUNTIME_OS_VERSIONING -DSWIFT_STDLIB_SHORT_MANGLING_LOOKUPS -DSWIFT_STDLIB_ENABLE_VECTOR_TYPES -DSWIFT_STDLIB_HAS_TYPE_PRINTING -DSWIFT_STDLIB_SUPPORTS_BACKTRACE_REPORTING -DSWIFT_STDLIB_ENABLE_UNICODE_DATA -DSWIFT_STDLIB_CONCURRENCY_TRACING -std=c++17 -fno-exceptions -fno-rtti -fno-exceptions -fno-rtti
1.	/Users/ktoso/code/swift-project/swift/stdlib/public/Concurrency/Task.cpp:1078:79: current parser token ';'
2.	/Users/ktoso/code/swift-project/swift/stdlib/public/Concurrency/Task.cpp:1067:53: parsing function body 'swift::swift_task_create'
3.	/Users/ktoso/code/swift-project/swift/stdlib/public/Concurrency/Task.cpp:1067:53: in compound statement ('{}')
4.	/Users/ktoso/code/swift-project/swift/stdlib/public/Concurrency/Task.cpp:1072:43: in compound statement ('{}')
5.	/Applications/XcodeSEIN.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/tuple:1182:12: instantiating class definition 'std::tuple<void (*&)(__attribute__((swift_async_context)) swift::AsyncContext *, void *)__attribute__((swiftasynccall)), unsigned long &>::_EnableAssignFromPair<false, const std::pair<void (*)(__attribute__((swift_async_context)) swift::AsyncContext *)__attribute__((swiftasynccall)), unsigned long> &>'
6.	/Applications/XcodeSEIN.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/tuple:1172:12: instantiating class definition 'std::tuple<void (*&)(__attribute__((swift_async_context)) swift::AsyncContext *, void *)__attribute__((swiftasynccall)), unsigned long &>::_AssignPredicateFromPair<std::is_assignable, false, const std::pair<void (*)(__attribute__((swift_async_context)) swift::AsyncContext *)__attribute__((swiftasynccall)), unsigned long> &>'
7.	/Applications/XcodeSEIN.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/usr/include/c++/v1/__type_traits/is_assignable.h:26:29: instantiating class definition 'std::is_assignable<void (*&)(__attribute__((swift_async_context)) swift::AsyncContext *, void *)__attribute__((swiftasynccall)), void (*const &)(__attribute__((swift_async_context)) swift::AsyncContext *)__attribute__((swiftasynccall))>'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  clang-13                 0x0000000105d7147c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  clang-13                 0x0000000105d7074c llvm::sys::RunSignalHandlers() + 112
2  clang-13                 0x0000000105d70c6c llvm::sys::CleanupOnSignal(unsigned long) + 236
3  clang-13                 0x0000000105ca9cac (anonymous namespace)::CrashRecoveryContextImpl::HandleCrash(int, unsigned long) + 100
4  clang-13                 0x0000000105ca9f38 CrashRecoverySignalHandler(int) + 124
5  libsystem_platform.dylib 0x00000001a03b6a84 _sigtramp + 56
6  libsystem_pthread.dylib  0x00000001a0387c28 pthread_kill + 288
7  libsystem_c.dylib        0x00000001a0295ae8 abort + 180
8  libsystem_c.dylib        0x00000001a0294e44 err + 0
9  clang-13                 0x00000001095ac5ec clang::ASTContext::mergeExtParameterInfo(clang::FunctionProtoType const*, clang::FunctionProtoType const*, bool&, bool&, llvm::SmallVectorImpl<clang::FunctionType::ExtParameterInfo>&) (.cold.4) + 0
10 clang-13                 0x0000000107ec8240 clang::ASTContext::mergeExtParameterInfo(clang::FunctionProtoType const*, clang::FunctionProtoType const*, bool&, bool&, llvm::SmallVectorImpl<clang::FunctionType::ExtParameterInfo>&) + 556
11 clang-13                 0x0000000107b68a5c clang::Sema::IsFunctionConversion(clang::QualType, clang::QualType, clang::QualType&) + 876
12 clang-13                 0x0000000107b8bd7c IsStandardConversion(clang::Sema&, clang::Expr*, clang::QualType, bool, clang::StandardConversionSequence&, bool, bool) + 1336
13 clang-13                 0x0000000107b6813c TryImplicitConversion(clang::Sema&, clang::Expr*, clang::QualType, bool, clang::Sema::AllowedExplicit, bool, bool, bool, bool) + 104
14 clang-13                 0x0000000107b6854c clang::Sema::PerformImplicitConversion(clang::Expr*, clang::QualType, clang::Sema::AssignmentAction, bool) + 272
15 clang-13                 0x00000001079123b4 clang::Sema::CheckSingleAssignmentConstraints(clang::QualType, clang::ActionResult<clang::Expr*, true>&, bool, bool, bool) + 904
16 clang-13                 0x000000010791cea4 clang::Sema::CheckAssignmentOperands(clang::Expr*, clang::ActionResult<clang::Expr*, true>&, clang::SourceLocation, clang::QualType, clang::BinaryOperatorKind) + 428
17 clang-13                 0x0000000107904f9c clang::Sema::CreateBuiltinBinOp(clang::SourceLocation, clang::BinaryOperatorKind, clang::Expr*, clang::Expr*) + 2264
```

File: stdlib/public/Concurrency/TaskGroup.cpp [hunk 6]
```diff
@@ -1291,6 +1290,14 @@ void DiscardingTaskGroup::offer(AsyncTask *completedTask, AsyncContext *context)
...
       // We must resume the waiting task with a success, so let us return here.
       resumeWaitingTask(completedTask, assumed, /*hadErrorResult=*/false, alreadyDecrementedStatus);
+
+//      // TODO: since the DiscardingTaskGroup ended up written as `-> T` in order to use the same pointer auth as the
+//      //       usual task closures; we end up retaining the value when it is returned. As this is a discarding group
+//      //       we actually can and should release this value.
+//      //       Is there a way we could avoid the retain made on the returned value entirely?
+//      if (completedTask->futureFragment()->getResultType()->isClassObject()) {
+//        swift_release(reinterpret_cast<HeapObject *>(completedTask->futureFragment()->getStoragePtr()));
+//      }
```
Comment (ktoso) [inline stdlib/public/Concurrency/TaskGroup.cpp:1300]: This is not what we want; instead we want to fix the signature of the group addTask

File: stdlib/public/Concurrency/TaskGroup.cpp [hunk 7]
```diff
@@ -1369,6 +1376,12 @@ void TaskGroupBase::resumeWaitingTask(
         // locks) because we know that the child task is completed and
         // we can't be holding its locks ourselves.
         _swift_taskGroup_detachChild(asAbstract(this), completedTask);
+        if (hadErrorResult) {
+          // We only used the task to keep the error in the future fragment around
+          // so now that we emitted the error and detached the task, we are free to release the task immediately.
+          auto error = reinterpret_cast<SwiftError *>(result.storage);
+          swift_release(completedTask); // we need to do this if the error is a class
+        }
```
Comment (ktoso) [inline stdlib/public/Concurrency/TaskGroup.cpp:1384]: This is somewhat weird; releasing the error does not do the right thing it seems (over releasing);
but releasing the task seems to be right since we retained this task to keep the error around. We could instead try to store only the error but that's a bigger refactoring.

**Comment by ktoso** [comment]
This is primarily stuck on pointer auth issues

File: include/swift/Runtime/Concurrency.h [hunk 8]
```diff
@@ -82,6 +82,11 @@ AsyncTaskAndContext swift_task_create(
 using FutureAsyncSignature =
   AsyncSignature<void(void*), /*throws*/ true>;
 
+/// Caution: not all functions actually throw, so
+/// this signature may be incorrect.
+using VoidAsyncSignature =
+  AsyncSignature<void(), /*throws*/ true>;
```
Comment (ktoso) [inline include/swift/Runtime/Concurrency.h:88]: AFAICS that's the right way to spell it

File: include/swift/ABI/MetadataValues.h [hunk 9]
```diff
@@ -1452,6 +1452,7 @@ namespace SpecialPointerAuthDiscriminators {
   const uint16_t EscalationNotificationFunction = 0x5be4; // = 23524
   const uint16_t AsyncThinNullaryFunction = 0x0f08; // = 3848
   const uint16_t AsyncFutureFunction = 0x720f; // = 29199
+  const uint16_t AsyncVoidFunction = 0x720f; // = FIXME!!!!!
```
[the remark this answers is not in the document]
Comment (al45tair) [reply include/swift/ABI/MetadataValues.h:1455]: I may be mistaken, but I think you actually want `AsyncThinNullaryFunction` here. As in, I told Swift to compile

```
func foo(bar: @Sendable @escaping () async -> Void) async {
  print("Calling")
  await bar()
  print("Called")
}
```

and it used `0x0f08` as the discriminator for the function `bar`.

[the remark this answers is not in the document]
Comment (ktoso) [reply include/swift/ABI/MetadataValues.h:1455]: Cool yeah I did not see it used but if it will do the right thing that'd be great!

Now just to resolve the compiler crashing... :) Thanks!

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: stdlib/public/Concurrency/TaskGroup.cpp [hunk 10]
```diff
@@ -1281,7 +1294,10 @@ void DiscardingTaskGroup::offer(AsyncTask *completedTask, AsyncContext *context)
           resumeWaitingTaskWithError(readyErrorItem.getRawError(this), assumed, alreadyDecrementedStatus);
           break;
         case ReadyStatus::Error:
-          resumeWaitingTask(readyErrorItem.getTask(), assumed, /*hadErrorResult=*/true, alreadyDecrementedStatus);
+          resumeWaitingTask(readyErrorItem.getTask(), assumed,
+                            /*hadErrorResult=*/true,
+                            alreadyDecrementedStatus,
+                            /*taskWasRetained=*/true);
```
Comment (ktoso) [inline stdlib/public/Concurrency/TaskGroup.cpp:1300]: hunting this one down took way too long... this is another case where we dequeue the retained error task and need to release it later

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: stdlib/public/Concurrency/Task.cpp [hunk 11]
```diff
@@ -1060,21 +1061,40 @@ void swift::swift_task_run_inline(OpaqueValue *result, void *closureAFP,
...
+    size_t rawTaskCreateFlags,
     TaskOptionRecord *options,
     const Metadata *futureResultType,
     void *closureEntry, HeapObject *closureContext) {
-  FutureAsyncSignature::FunctionType *taskEntry;
-  size_t initialContextSize;
-  std::tie(taskEntry, initialContextSize) =
+  TaskCreateFlags taskCreateFlags(rawTaskCreateFlags);
+
+  if (taskCreateFlags.isDiscardingTask()) {
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.cpp:1070]: the pointer auth dance for the nullary function type:

**Review by rjmccall — APPROVED** [review]
Okay.  I think this seems reasonable.

**Comment by ktoso** [comment]
A bit confused why same commits worked on Windows 5.9 but not here, in a lit configuration failure. Let's try again

**Comment by ktoso** [comment]
@swift-ci please smoke test Windows

**Comment by ktoso** [comment]
Silly main + windows only issue in lit config:

```
  File "C:/Users/swift-ci/jenkins/workspace/swift-PR-windows/swift\test\lit.cfg", line 2324, in <module>
    escape_for_substitute_captures(config.target_run_with_leaks))
AttributeError: 'TestingConfig' object has no attribute 'target_run_with_leaks'
```

Fixed I hope;

@swift-ci please smoke test and merge

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

**Comment by ktoso** [comment]
Pending on the CI unbreak: https://github.com/apple/swift/pull/66138

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge
